### ISSUE TITLE
[Proposal] Extend with operator to anonymous type

### ISSUE THREAD

**Issue by leandromoh** [dotnet/csharplang#3530]
The `with` expression, introduced in C# 9, is used to create new object that’s a copy of a immutable one, but changing the value of some members. 

Since anonymous types are also immutable, should be nice to use `with` expression with them too.

Proposed:

```c#
var person = new { FirstName = "Scott", LastName = "Hunter", Age = 25 };

var otherPerson = person with { LastName = "Hanselman" };
```

Currently:

```c#
var person = new { FirstName = "Scott", LastName = "Hunter", Age = 25 };

var otherPerson = new { person.FirstName, LastName = "Hanselman", person.Age };
```

**Comment by leandromoh** [comment]
In case of someone is looking for such feature, here is [a gist](https://gist.github.com/leandromoh/d81d8ad17f34d10c5b5d0ed5ac13fe87) with my workaround. with follow usage:
```c#
var otherPerson = person.With(new { LastName = "Hanselman" });
```

**Comment by 333fred** [comment]
I'm willing to champion this extension, with the caveat that it won't come in 9 :)

**Comment by HaloFour** [comment]
Makes sense, you could probably consider anonymous types to automatically be records given they already have value equality, no?

**Comment by 333fred** [comment]
Possibly. It would still depend on whether we add anything else to the definition of a record type. I'm not sure whether we'd add a copy constructor, for example, or if we'd just do the copy at call site.

**Comment by 333fred** [comment]
@leandromoh this will need an actual spec. Would you be willing to take a stab at writing one? It would need to be an extension to the existing records proposal.

**Comment by HaloFour** [comment]
Would "withers" come for free if an anonymous type definition was modified so that:

1. Every property was `{ get; init; }` instead of `{ get; }`, and
1. The anonymous type exposed a `Clone` method which would create a shallow clone of the anonymous type?

**Comment by 333fred** [comment]
Probably, but then every piece of code generates an anonymous type would need to be recompiled to work with this feature.

**Comment by svick** [comment]
@333fred Since anonymous types can't be t exposed publicly, why would that be an issue?

**Comment by 333fred** [comment]
I suppose it might be fine. Still, we need a spec to actually evaluate this :)

**Comment by leandromoh** [comment]
> @leandromoh this will need an actual spec. Would you be willing to take a stab at writing one? It would need to be an extension to the existing records proposal.

@333fred I would like to do it, of course. I just need to know exactly what and how to do, I will be glad to contribute more.

**Comment by 333fred** [comment]
Basically, look at the existing records proposal (https://github.com/dotnet/csharplang/blob/master/proposals/records.md#with-expression) and detail exactly how it would need to be modified in order to support `with`ing an anonymous type. How would it actually function? Would we make changes to how anonymous types are emitted as @HaloFour suggested to generate a copy constructor & mark the properties as `init`, or would we do codegen at the `with` invocation site? I think the former approach has promise, but it needs to be investigated and formalized. The proposal template is here: https://github.com/dotnet/csharplang/blob/master/proposals/proposal-template.md. Note: don't make a PR yet, just filling it out in this issue is fine for now.

**Comment by HaloFour** [comment]
Given that the compiler has complete control over the definition and use of anonymous types it might make more sense to have the compiler treat it special and just emit its own call to the constructor manually copying the source properties, as suggested in the original post.  I don't see any other added benefit to adding/changing the members of the anonymous type, it's not like external code can take advantage of it outside of reflection.

**Comment by ChayimFriedman2** [comment]
Will it allow adding properties, since anonymous types do not have declared type like records?

**Comment by HaloFour** [comment]
@ChayimFriedman2 

That sounds interesting, although I'd a little concerned that it would make it very easily to make a mistake:

```cs
var person1 = new { Name = "Bill", Age = 45 };
var person2 = person1 with { AGe = 46 }; // oops!
```

**Comment by ChayimFriedman2** [comment]
Right. That seems like VB's implicit variable declaration, which is known to be error-prone...

**Comment by leandromoh** [comment]
@333fred @HaloFour proposal added in the description of this issue. Feel free to edit it.

**Comment by AartBluestoke** [comment]
@HaloFour perhaps 'anon with' {x=y} requires x to be an already existing variable (ie, you can't extend an anon type with 'with')

**Comment by 333fred** [comment]
@leandromoh it's a good start, but it needs to be more concrete. A spec isn't a guess about how to implement something: it lays out exactly how it will be implemented.

**Comment by leandromoh** [comment]
@333fred okay, I will try make it more detailed, but probably it will need some review after all.

**Comment by leandromoh** [comment]
@ChayimFriedman2 @HaloFour javascript has a [spread operator](https://medium.com/@mwalterspieler_31762/es6-using-the-spread-operator-to-merge-object-d71c62c98056) that makes exactly this. example:

```js
var obj1 = { name: "bob",  type:"Apple" };
var obj2 = { name: "joe", price: 0.20 }; 
var merge = {...obj1, ...obj2, age: 34 }; 

console.log(merge); 

console.log(merge); 
/* { 
   age: 34
   name: "joe"
   price: 0.2
   type: "Apple"
}; */
```

**Comment by leandromoh** [comment]
@333fred "Detailed design" section updated in the description. Hope it is closer of what is need.

**Comment by leandromoh** [comment]
For note: [with expressions on non records on LDM-2020-06-22](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-22.md#with-expressions-on-non-records)

**Comment by MgSam** [comment]
If this were pre-C#-7.0 I would think this was a great idea. Post-C#-7.0 I don't see any point of putting development work into anonymous types. Tuples are better in nearly every respect - they are structs, they can be return types and parameter types, and if you really really need to mutate them, you already can.

Given that, how often, and why, do you really need to "mutate" an anonymous type? The use cases for this seem vanishingly small.

**Comment by leandromoh** [comment]
@MgSam this proposal enables anonymous types practically act like anonymous record types, as they have the rest of the properties of a record already: value equality, ToString, etc.

**Comment by JKamsker** [comment]
It would be also cool if those ``with`` operators could extend the anonymous type. 
My usecase:

```csharp
var files = Directory.EnumerateFiles(dir, "*.*", SearchOption.TopDirectoryOnly)
  .Select(x => new FileInfo(x))
  .Select(x => new
  {
      File = x,
      FileName = x.Name,
      Match = myRegex.Match(x.Name)
  })
  .Where(x => x.Match.Success)
  .Select(x => x with (new
  {
      Group = x.Match.Groups[1].Value,
      FileNumber = x.Match.Groups[2].Value,
  }));
```

The Resulting anonymous type should then contain

```
File
FileName
Match
Group
FileNumber
```

**Comment by darkflame0** [comment]
> It would be also cool if those `with` operators could extend the anonymous type. My usecase:
> 
> ```cs
> var files = Directory.EnumerateFiles(dir, "*.*", SearchOption.TopDirectoryOnly)
>   .Select(x => new FileInfo(x))
>   .Select(x => new
>   {
>       File = x,
>       FileName = x.Name,
>       Match = myRegex.Match(x.Name)
>   })
>   .Where(x => x.Match.Success)
>   .Select(x => x with
>   {
>       Group = x.Match.Groups[1].Value,
>       FileNumber = x.Match.Groups[2].Value
>   }));
> ```
> 
> The Resulting anonymous type should then contain
> 
> ```
> File
> FileName
> Match
> Group
> FileNumber
> ```
any progress on this?

my use case is flattening IGrouping.
```cs
var group = data.GroupBy(a=>a.KeyProp);
var flat = group.SelectMany(a=>a.Select(b=>b with {Key = a.KeyProp}));
```

**Comment by darkflame0** [comment]
but using with expression for extended objects can be misleading.
meybe this is better?

> @ChayimFriedman2 @HaloFour javascript has a [spread operator](https://medium.com/@mwalterspieler_31762/es6-using-the-spread-operator-to-merge-object-d71c62c98056) that makes exactly this. example:
> 
> ```js
> var obj1 = { name: "bob",  type:"Apple" };
> var obj2 = { name: "joe", price: 0.20 }; 
> var merge = {...obj1, ...obj2, age: 34 }; 
> 
> console.log(merge); 
> /* { 
>    age: 34
>    name: "joe"
>    price: 0.2
>    type: "Apple"
> }; */
> ```
> 
> If objects have common properties the priority will be right to left. The object the most at the right will have priority over the one at its left and so on. If we could expand C#'s anonymous type in a similar way would be amazing.


```cs
var flat = group.SelectMany(a=>a.Items.Select(b=> {Key = a.KeyProp, ...b}));
```

**Comment by qrli** [comment]
Introducing spread operator would be much bigger feature. And there would be complicated cases where `obj1` and `obj2` have properties with the same name but different types, for example.

Here using `with` is a natural extension of an existing feature, with well defined behavior.

**Comment by mrwensveen** [comment]
> @ChayimFriedman2
> 
> That sounds interesting, although I'd a little concerned that it would make it very easily to make a mistake:
> 
> ```cs
> var person1 = new { Name = "Bill", Age = 45 };
> var person2 = person1 with { AGe = 46 }; // oops!
> ```

On the other hand, the analogous F# feature does allow exactly this on anonymous record types.

**Comment by JKamsker** [comment]
> > @ChayimFriedman2
> > 
> > That sounds interesting, although I'd a little concerned that it would make it very easily to make a mistake:
> > 
> > ```cs
> > var person1 = new { Name = "Bill", Age = 45 };
> > var person2 = person1 with { AGe = 46 }; // oops!
> > ```
> 
> On the other hand, the analogous F# feature does allow exactly this on anonymous record types.

Such errors can be avoided by using intellisense, which should know existing anonymous properties.
